build_runner/lib/src/build_script_generate/build_script_generate.dart

```diff
@@ -109,11 +133,71 @@ Future<Iterable<Expression>> findBuilderApplications({
       .expand((c) => c.postProcessBuilderDefinitions.values)
       .where(_isValidDefinition);
 
-  return [
+  final applications = [
     for (var builder in orderedBuilders) _applyBuilder(builder),
     for (var builder in postProcessBuilderDefinitions)
       _applyPostProcessBuilder(builder)
   ];
+
+  final canRunWithSoundNullSafety = await _allMigratedToNullSafety(
+      packageGraph,
+      reader,
+      orderedBuilders
+          .map((e) => e.import)
+          .followedBy(postProcessBuilderDefinitions.map((e) => e.import)));
+
+  return _BuildScriptInfo(applications, canRunWithSoundNullSafety);
+}
+
+Future<bool> _allMigratedToNullSafety(PackageGraph packageGraph,
+    AssetReader reader, Iterable<String> imports) async {
+  final baseForRelative = AssetId(packageGraph.root.name, scriptLocation);
+
+  for (final import in imports.toSet()) {
+    final id = AssetId.resolve(Uri.parse(import), from: baseForRelative);
+    String content;
+    try {
+      content = await reader.readAsString(id);
+    } on AssetNotFoundException {
+      // Build is likely to be invalid, so the language version shouldn't
+      // matter. Still, don't assume null safety.
+      return false;
+    }
+
+    final rawVersion = packageGraph.allPackages[id.package]?.languageVersion;
+    if (rawVersion == null) {
+      // Can't determine language version of import, bail out
+      return false;
+    }
+    var version = Version(rawVersion.major, rawVersion.minor, 0);
+    final parsedFile = parseString(
+      content: content,
+      featureSet: FeatureSet.fromEnableFlags2(
+          sdkLanguageVersion: version, flags: const []),
+      throwIfDiagnostics: false,
+    );
+    if (parsedFile.errors.isNotEmpty) {
+      // Don't try to assume a language version if the imported file is invalid
+      return false;
+    }
+
+    final definedLanguageVersion = parsedFile.unit.languageVersionToken;
```

Comment:
Instead of manually doing this detection based on versions, we should be able to do something like:

```
return parsedFile.unit.featureSet.isEnabled(Feature.non_nullable);
```